core/src/main/java/hudson/slaves/SlaveComputer.java

```diff
@@ -1022,7 +1022,7 @@ public Void call() {
             SLAVE_LOG_HANDLER = new RingBufferLogHandler(ringBufferSize) {
                 Formatter dummy = new SimpleFormatter();
                 @Override
-                public synchronized void publish(LogRecord record) {
+                public /* not synchronized */ void publish(LogRecord record) {
```

Comment:
Note that the `super` call [is still synchronized](https://stackoverflow.com/a/15374546/12916).

Reply:
I am not really sure this will solve the problem, because, as you say, the super is still synchronized. The deadlock is caused by both the RemoteClassLoader that is involved by loading the new overloaded RingBufferLogHandler and de publish synchronized call. Perhaps if a separate  new FormattedRingBufferLogHandler was loaded earlier, then the lock RemoteClassLoader would not be at the same time as the publish synchronized lock.

Reply:
AFAICT from the thread dump the deadlock is caused by the following line of code forcing some class loading, so this should bypass the issue. There is not enough information in a thread dump to tell which class needed to be loaded and thus what the real problem was.

Reply:
To add a bit to that: https://github.com/jenkinsci/jenkins/blob/6ede8dbe3a290b70e958624e01549debe24baaed/core/src/main/java/hudson/slaves/SlaveComputer.java#L1026 is apparently triggering class loading

```
"Channel reader thread: channel":
  waiting to lock monitor 0x00007fbd9401aed8 (object 0x000000076ded1cc8, a hudson.remoting.RemoteClassLoader),
…
"Channel reader thread: channel":
	at hudson.slaves.SlaveComputer$SlaveInitializer$1.publish(SlaveComputer.java:1026)
```

but the JVM thread dump does not indicate what the initiating class loader was, what class name is being loaded, which subexpression is forcing class loading, etc. In non-deadlock situations, IIRC, you would see a call stack frame from a `ClassLoader.loadClass` override directly above the familiar source code line, as if it were loading classes reflectively.

Reply:
I think that the class loaded is the dynamically overrided RingBufferLogHandler or only the new overrided publish method for this class. It could very well be that this will indeed fix the problem because now synchronized for the publish is only in the parent class , and that is already loaded, and it does not need to load and synchronize at the same time. So, lets hope that this will work and thanks for your  efforts.

Reply:
> I think that the class loaded is the dynamically overrided RingBufferLogHandler

`SlaveComputer$SlaveInitializer$1`? But that is clearly already loaded, because we are running a method on it.

It is possible `Stream` or similar is being loaded for the first time. That is the sort of weird problem I recall from the code link in the PR description.

Initially I thought that it might be some class linked to one of the record parameters—not `p.getClass()` itself, which is clearly loaded since we have created the parameter, but a dependency of that parameter class needed to run some methods on it such as `toString()`. But that cannot be right because in the line 1026 mentioned in the report we are not calling any methods on `p`; that would come in the next line https://github.com/jenkinsci/jenkins/blob/6ede8dbe3a290b70e958624e01549debe24baaed/core/src/main/java/hudson/slaves/SlaveComputer.java#L1027 and we would expect to see some more stack frames where `SimpleFormatter.formatMessage` is being called and then calling `SomeRemotelyLoadedPluginType.toString`.